pkg/traceql/engine_metrics_average.go

```diff
@@ -0,0 +1,540 @@
+
+func newAverageOverTimeMetricsAggregator(attr Attribute, by []Attribute) *AverageOverTimeAggregator {
+	return &AverageOverTimeAggregator{
+		attr: attr,
+		by:   by,
+	}
+}
+
+func (a *AverageOverTimeAggregator) init(q *tempopb.QueryRangeRequest, mode AggregateMode) {
+	exemplarFn := func(s Span) (float64, uint64) {
+		return math.NaN(), a.spanStartTimeMs(s)
+	}
+
+	a.seriesAgg = &SpanSetsAverageOverTimeAggregator{
+		ss:              make(SeriesSet),
+		countProm:       make(map[string]string),
+		len:             IntervalCount(q.Start, q.End, q.Step),
+		start:           q.Start,
+		end:             q.End,
+		step:            q.Step,
+		exemplarBuckets: newBucketSet(IntervalCount(q.Start, q.End, q.Step)),
+	}
+
+	if mode == AggregateModeRaw {
+		a.agg = NewAvgOverTimeSpanAggregator(a.attr, a.by, q.Start, q.End, q.Step)
+	}
+
+	a.exemplarFn = exemplarFn
+	a.mode = mode
+}
+
+func (a *AverageOverTimeAggregator) observe(span Span) {
+	a.agg.Observe(span)
+}
+
+func (a *AverageOverTimeAggregator) observeExemplar(span Span) {
+	v, ts := a.exemplarFn(span)
+	a.agg.ObserveExemplar(span, v, ts)
+}
+
+func (a *AverageOverTimeAggregator) observeSeries(ss []*tempopb.TimeSeries) {
+	a.seriesAgg.Combine(ss)
+}
+
+func (a *AverageOverTimeAggregator) result() SeriesSet {
+	if a.agg != nil {
+		return a.agg.Series()
+	}
+
+	// In the frontend-version the results come from
+	// the job-level aggregator
+	ss := a.seriesAgg.Results()
+	if a.mode == AggregateModeFinal {
+		for i := range ss {
+			if strings.Contains(i, "_type") {
+				delete(ss, i)
+			}
+		}
+	}
+	return ss
+}
+
+func (a *AverageOverTimeAggregator) extractConditions(request *FetchSpansRequest) {
+	// For metrics aggregators based on a span attribute we have to include it
+	includeAttribute := a.attr != (Attribute{}) && !request.HasAttribute(a.attr)
+	if includeAttribute {
+		request.SecondPassConditions = append(request.SecondPassConditions, Condition{
+			Attribute: a.attr,
+		})
+	}
+
+	for _, b := range a.by {
+		if !request.HasAttribute(b) {
+			request.SecondPassConditions = append(request.SecondPassConditions, Condition{
+				Attribute: b,
+			})
+		}
+	}
+}
+
+func (a *AverageOverTimeAggregator) validate() error {
+	if len(a.by) >= maxGroupBys {
+		return newUnsupportedError(fmt.Sprintf("metrics group by %v values", len(a.by)))
+	}
+	return nil
+}
+
+func (a *AverageOverTimeAggregator) spanStartTimeMs(s Span) uint64 {
+	return s.StartTimeUnixNanos() / uint64(time.Millisecond)
+}
+
+func (a *AverageOverTimeAggregator) String() string {
+	s := strings.Builder{}
+
+	s.WriteString(metricsAggregateAvgOverTime.String())
+	s.WriteString("(")
+	if a.attr != (Attribute{}) {
+		s.WriteString(a.attr.String())
+	}
+	s.WriteString(")")
+
+	if len(a.by) > 0 {
+		s.WriteString("by(")
+		for i, b := range a.by {
+			s.WriteString(b.String())
+			if i < len(a.by)-1 {
+				s.WriteString(",")
+			}
+		}
+		s.WriteString(")")
+	}
+	return s.String()
+}
+
+type SpanSetsAverageOverTimeAggregator struct {
```

Comment:
nit:  I would rename this to `averageOverTimeSeriesAggregator`.  Spansets is something else in traceql, a collection of individual spans. This is combining time series, so think this name fits better. 

Related:   I don't think these structs are used outside of this package, so let's unexport them by lower casing.